SparseGrids/tsgRuleLocalPolynomial.hpp

```diff
@@ -364,6 +463,79 @@ class templRuleLocalPolynomial : public BaseRuleLocalPolynomial{
         }
         return value;
     }
+
+    // Same logic as the corresponding eval functions.
+    double diffPWQuadratic(int point, double x) const{
+        if (rule == rule_localp) {
+            if (point == 1) return -1.0;
+            if (point == 2) return 1.0;
+        } else if (rule == rule_localpb) {
+            if (point == 0) return -1.0;
+            if (point == 1) return 1.0;
+        }
+        return -2.0 * x;
+    }
+    double diffPWCubic(int point, double x) const{
+        if (rule == rule_localp) {
+            if (point == 0) return 0.0;
+            if (point == 1) return -1.0;
+            if (point == 2) return 1.0;
+            if (point <= 4) return -2.0 * x;
+        } else if (rule == rule_localpb) {
+            if (point == 0) return -1.0;
+            if (point == 1) return 1.0;
+            if (point == 2) return -2.0 * x;
+        } else if (rule == rule_localp0) {
+            if (point == 0) return -2.0 * x;
+        }
+        return (point % 2 == 0) ? 1.0 / 3.0 - x * (x + 2.0) : -1.0 / 3.0 + x * (x - 2.0);
+    }
+    double diffPWPower(int point, double x) const{
+        if (rule == rule_localp)     if (point <= 8) return diffPWCubic(point, x);
+        if (rule == rule_semilocalp) if (point <= 4) return diffPWCubic(point, x);
+        if (rule == rule_localpb)    if (point <= 4) return diffPWCubic(point, x);
+        if (rule == rule_localp0)    if (point <= 2) return diffPWCubic(point, x);
+        int level = getLevel(point);
+        int max_ancestors;
+        if (rule == rule_localp)     max_ancestors = level-2;
+        if (rule == rule_semilocalp) max_ancestors = level-1;
+        if (rule == rule_localpb)    max_ancestors = level-1;
+        if (rule == rule_localp0)    max_ancestors = level;
+        if (max_order > 0) max_ancestors = std::min(max_ancestors, max_order - 2);
+        int most_turns = 1;
```

Comment:
Add a new line above this, since this is code with separate logic, makes it easier to see the logic of the lambda.

Comment on what the lambda does, e.g., captures by reference `most_turns` and `phantom_distance`, uses those as internal state variables, on each call it returns the next normalized ancestor node.

Reply:
Changed.